docs/guide/cpp-state/README.md

```diff
@@ -0,0 +1,163 @@
+---
+title: Dealing with State in C++
+---
+
+# Dealing with State in C++
+
+Nodes outputs of which depend _solely_ on their inputs in at any point of time
+are cool. They easy to understand, test, and compose. But building a useful
+device using only such pure nodes is not realistic. Someone needs to keep state
+along program run time.
+
+A node can define state data that will persist for the time the program
+executes. In other words, a node can put some value to the state in
+`evaluate` call and retrieve that value in any of subsequent `evaluate`
+invocations.
+
+## The task
+
+Let’s make a simple `count` node that will increment a value by one each time a
+pulse is sent to it. Also, we’ll make the step size configurable and provide a
+pulse input to reset the counter to zero.
+
+<div class="ui segment note">
+<span class="ui ribbon label">Note</span>
+XOD standard library already includes <a
+href="/libs/xod/core/count/"><code>xod/core/count</code></a>
+node with similar functionality. But let’s forget about it for a moment.
+</div>
+
+## Prepare the node
+
+As always, when you [make a C\++ node](../nodes-for-xod-in-cpp), start with a
+new patch, add required terminals, and the `not-implemented-in-xod` node.
+
+![Patch outline](./outline.patch.png)
+
+It’s a good idea to provide a resonable default value for `STEP`. We’ll set
+it to 1.
+
+Double-click on `not-implemented-in-xod` node to open the code editor.
+
+## Define a state shape
+
+You define the persistent state using the `State` struct in C\++. In our case,
+we need to store a single counter value, so our struct will have a single field.
+Let’s call it `counterValue`:
+
+```cpp
+struct State {
+    Number counterValue = 0;
+};
+```
+
+All state value regardless of type start with their default values. The default
+value of numbers is `0`, so this assignment is not required. We add it here just
+to demonstrate a possibility to initialize with another initial value.
```

Comment:
It's not clear, If I need default `0` value, should I specify `Number counterValue` or not.
On the other hand, the article is for advanced guys who may do something with C++, and perhaps this remark is not actual.

Reply:
Not sure what do you mean. Isn’t “*so this assignment is not required*” says it is optional? Or do you talk about another thing?

Reply:
I mean could User just leave State empty, or he still need to write `Number counterValue` but without value ;)

Reply:
Ahem… he uses `counterValue`, so he definitely required to declare it. OK, I’ve got the problem, will clarify it a bit.